dr_backup/dr_backup.py

```diff
@@ -0,0 +1,304 @@
+    return {k: v for k, v in kwargs.items() if v is not None}
+
+
+def assume_session(
+        role_session_name: str,
+        role_arn: str,
+        region_name: str,
+        duration_seconds: int = None,
+) -> boto3.Session:
+    """
+    Returns a session with the given name and role.
+    If not specified, duration will be set by AWS, probably at 1 hour.
+    Region can be overridden by each client or resource spawned from this session.
+    """
+    assume_role_kwargs = filter_none_values(
+        {
+            "RoleSessionName": role_session_name,
+            "RoleArn": role_arn,
+            "DurationSeconds": duration_seconds,
+        }
+    )
+    credentials = boto3.client("sts").assume_role(**assume_role_kwargs)["Credentials"]
+    create_session_kwargs = filter_none_values(
+        {
+            "aws_access_key_id": credentials["AccessKeyId"],
+            "aws_secret_access_key": credentials["SecretAccessKey"],
+            "aws_session_token": credentials["SessionToken"],
+            "region_name": region_name,
+        }
+    )
+    print(f"Assumed role {role_arn}")
+    return boto3.Session(**create_session_kwargs)
+
+
+def main():
+    aws_region = "eu-west-1"
+    cluster_id = os.environ["DB_ID"]
+    source_account = os.environ["SOURCE_ACCOUNT"]
+    kms_key_id = os.environ["KMS_KEY_ID"]
+    shared_kms_key_id = f"arn:aws:kms:{aws_region}:{source_account}:key/{kms_key_id}"
+    backup_account = os.environ["BACKUP_ACCOUNT"]
+    backup_acc_role_arn = os.environ["BACKUP_ACCOUNT_ROLE"]
+    backups_to_keep = 7200
+
+    client = boto3.client('rds', region_name=aws_region)
+
+    print(f"Database to backup {cluster_id}")
+
+    snapshot_identifier = get_latest_snapshot(client, cluster_id)
+    target_snapshot_identifier = snapshot_identifier.replace("rds:", "")
+    shared_snapshot_identifier = f"arn:aws:rds:{aws_region}:{source_account}:cluster-snapshot:{target_snapshot_identifier}"
+    backup_snapshot_identifier = f"{target_snapshot_identifier}-bck"
+
+    # Delete target if it exists
+    delete_snapshot(client, target_snapshot_identifier)
+
+    # Copy snapshot over to manual using KMS key
+    copy_latest_snapshot(client, snapshot_identifier, target_snapshot_identifier, kms_key_id)
+
+    # Share the snapshot with sandbox
+    share_snapshot(client, target_snapshot_identifier, backup_account)
+
+    backup_session = assume_session(
+        'backup_session',
+        backup_acc_role_arn,
+        aws_region)
+
+    backup_client = backup_session.client('rds', region_name=aws_region)
+
+    # Delete backup acc target if it exists
+    delete_snapshot(backup_client, backup_snapshot_identifier)
+
+    copy_individual_snapshot(
+        backup_client,
+        shared_snapshot_identifier,
+        backup_snapshot_identifier,
+        shared_kms_key_id,
+        aws_region
+    )
+
+    # Delete shared manual backup
+    delete_snapshot(client, target_snapshot_identifier)
```

Comment:
Not sure I've followed all of the above but is there any way we could get the wrong snapshot here?

Reply:
Delete snapshot only deletes from manual snapshot section so not the automated ones and it has a different identifier. It should be fine.